### PR TITLE
Task/jenkins 37245 Developer can jump to the run history for a specific branch and pull request

### PR DESCRIPTION
**Pull request by dragoonis**
# Description

See https://issues.jenkins-ci.org/browse/JENKINS-37245

Can jump to history for a branch

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by i386** [comment]
Gave this a go - LGTM :)

**Comment by michaelneale** [comment]
yes works for me - but I would like @kzantow @imeredith or @cliffmeyers to give it a quick look so we can get this in to next release. 

Functionally done (and I checked with editor with extra icon, and it is great) - just need some js devs to review!

nice one @dragoonis !

File: bin/watch-css.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -xe
+
+DIR="$( cd "$( dirname "$0" )" && pwd )"
+
+cd ${DIR}/../blueocean-web
+npm install ../blueocean-core-js
+npm install -g slink
```
Comment (michaelneale) [inline bin/watch-css.sh]: can it be a local install for this to work?

**Comment by sophistifunk** [comment]
Just the one issue I see with the shell script, but the JS code looks fine to me. 👍

File: bin/watch-css.sh (hunk 1 above)
Comment (kzantow) [inline bin/watch-css.sh]: All the docs refer to slink as being used globally, but we could (and should) make sure it doesn't have to be

File: blueocean-dashboard/src/main/less/core.less [hunk 2]
```diff
@@ -64,7 +64,7 @@
     margin-top: 10px;
 
     th {
-        width: 75px;
+        width: 100px;
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/less/core.less]: Should every column be pushed to 100px, or just the one that needs it? Could be that all the other columns are already getting explicit widths down below... although if we want to size the "actions" column explicitly, it might be good to slap a `className` on that one.

File: blueocean-dashboard/src/main/js/components/Branches.jsx [hunk 3]
```diff
@@ -86,12 +91,20 @@ export default class Branches extends Component {
...
+
                     <RunButton
                       className="icon-button"
                       runnable={branch}
                       latestRun={branch.latestRun}
                       onNavigation={openRunDetails}
                     />
+
+                    <div className="history-button-component">
+                        <Link to={historyButtonUrl} className="materials-icons">
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: Is this class name being used?

File: blueocean-dashboard/src/main/js/components/PullRequest.jsx [hunk 4]
```diff
@@ -72,12 +74,20 @@ export default class PullRequest extends Component {
...
+
                     <RunButton
                       className="icon-button"
                       runnable={pr}
                       latestRun={pr.latestRun}
                       onNavigation={openRunDetails}
                     />
+
+                    <div className="history-button-component">
+                        <Link to={historyButtonUrl} className="materials-icons">
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/PullRequest.jsx]: Is this class name being used?

File: blueocean-dashboard/src/main/less/history-button.less [hunk 5]
```diff
@@ -0,0 +1,39 @@
+.history-button-component {
+
+    display: inline-block;
+
+    .run-button, .stop-button {
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/less/history-button.less]: Are these classes relevant to the history button?

**Comment by cliffmeyers** [comment]
@dragoonis in future PR's, please create a branch name in the form `task/JENKINS-12345-some-short-descriptive-name` - makes it easier to find the branch in the list, and will also give your PR a descriptive name by default. Thanks :)

File: blueocean-dashboard/src/main/less/core.less (hunk 2 above)
Comment (sophistifunk) [inline blueocean-dashboard/src/main/less/core.less]: This will do for now. When we switch the table layouts over (hopefully this week) we'll need to take the number of extensions into account when sizing action-button-y columns. Luckily, the (baseline) column widths are runtime, not CSS in the new flexbox-based system.

File: blueocean-dashboard/src/main/js/components/Branches.jsx (hunk 3 above)
Comment (michaelneale) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: it's not terrible to have named divs in some cases (I know I often look for them, saves a lot of time in debugging and testing - which isn't a good reason, but I believe its not terrible practice as well).

File: blueocean-dashboard/src/main/less/history-button.less (hunk 5 above)
Comment (michaelneale) [inline blueocean-dashboard/src/main/less/history-button.less]: mayy this be a copy-paste error?

**Comment by michaelneale** [comment]
So in summary - looks like this is close: 

To merge: 
* Get rid of global install of slink
* Cliff had a Q above about a "relevant to the history button" style, may be in error

I believe other Q's have been addressed.

**Comment by dragoonis** [comment]
@i386 @cliffmeyers @sophistifunk  @michaelneale i've removed the shell script from this PR and can deal with it separately.

File: blueocean-dashboard/src/main/less/history-button.less (hunk 5 above)
Comment (dragoonis) [inline blueocean-dashboard/src/main/less/history-button.less]: I copied from the run-button.less .. I'l get rid of this.

**Comment by dragoonis** [comment]
CSS has been improved / sorted.

Next branch I make will have a descriptive name.

File: blueocean-dashboard/src/main/js/components/Branches.jsx [hunk 6]
```diff
@@ -58,6 +62,7 @@ export default class Branches extends Component {
         }
         const cleanBranchName = decodeURIComponent(branch.name);
         const runDetailsUrl = buildRunDetailsUrl(branch.organization, pipeline.fullName, cleanBranchName, latestRun.id, 'pipeline');
+        const historyButtonUrl = `${buildPipelineUrl(branch.organization, pipeline.fullName)}/activity/${branch.name}`;
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: Hey, this needs updating to use a query parameter instead... minor change, should be `${buildPipelineUrl(branch.organization, pipeline.fullName)}/activity?branch=${branch.name}` I think

File: blueocean-dashboard/src/main/js/components/PullRequest.jsx [hunk 7]
```diff
@@ -47,7 +49,7 @@ export default class PullRequest extends Component {
         const result = latestRun.result === 'UNKNOWN' ? latestRun.state : latestRun.result;
         const { fullName, organization } = contextPipeline;
         const runDetailsUrl = buildRunDetailsUrl(organization, fullName, decodeURIComponent(latestRun.pipeline), latestRun.id, 'pipeline');
-
+        const historyButtonUrl = `${buildPipelineUrl(organization, contextPipeline.name)}/activity/${pr.name}`;
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/PullRequest.jsx]: Same thing here

File: blueocean-dashboard/src/main/js/components/Branches.jsx [hunk 8]
```diff
@@ -58,6 +62,7 @@ export default class Branches extends Component {
         }
         const cleanBranchName = decodeURIComponent(branch.name);
         const runDetailsUrl = buildRunDetailsUrl(branch.organization, pipeline.fullName, cleanBranchName, latestRun.id, 'pipeline');
+        const historyButtonUrl = `${buildPipelineUrl(branch.organization, pipeline.fullName)}/activity?branch=${branch.name}`;
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: As noted in the previous comment, I think this needs to be:
`${buildPipelineUrl(branch.organization, pipeline.fullName)}/activity?branch=${encodeURIComponent(branch.name)}`;

Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: I guess it matters less since it's a query parameter...

**Comment by michaelneale** [comment]
@dragoonis mind if I get this up to date with master? 

Note to self: need to change funky branch names

**Comment by dragoonis** [comment]
Fire in Michael. Going to tweak as per Keith's suggestion.

**Comment by michaelneale** [comment]
ATH run: https://ci.blueocean.io/job/ATH-Jenkinsfile/job/master/1955/

**Comment by michaelneale** [comment]
ok have updated to master, looking good so far. Small tweaks as per Keith incoming I presume?

**Comment by i386** [comment]
Would be cool to get this into 1.1!

**Comment by cliffmeyers** [comment]
I'll plan to grab this and finish it off next week, unless we hear back from @dragoonis first.

File: blueocean-dashboard/src/main/js/components/Branches.jsx (hunk 8 above)
Comment (i386) [inline blueocean-dashboard/src/main/js/components/Branches.jsx]: Fixed

File: blueocean-dashboard/src/main/less/variables.less [hunk 9]
```diff
@@ -12,3 +12,7 @@
 @step-input-text-color: #4a4a4a;
 
 @errorscreen-bg-color: #949393;
+
+
+@brand-primary:         #4A90E2;
```
Comment (kzantow) [inline blueocean-dashboard/src/main/less/variables.less]: y u no have access to jdl ;)

**Comment by michaelneale** [comment]
ok works for me.

**Comment by i386** [comment]
Thanks for this one @dragoonis 👍

**Comment by dragoonis** [comment]
@i386 no problem. I've had to take a break for a while but I'm about to return again :) /cc @michaelneale
